### ISSUE TITLE
Elastic agent gets stuck trying to install application

### ISSUE THREAD

**Issue by AndersonQ** [elastic/elastic-agent#191]
If the elastic agent finds the application to be installed on its `download` folder, it'll unpack and install. However if the install operation fails, it'll retry it indefinitely, preventing any other application to be installed. Others could succeed.

Here are some debug logs showing it stuck on the retry while trying to install the elastic-endpoint.
![agent](https://user-images.githubusercontent.com/203669/157695467-042bd238-a8e6-40f1-a68c-e68ad04fbaef.png)

**Comment by jlind23** [comment]
@nimarezainia 
What should the behaviour to observe here? Should we try only a couple of times until moving to the next step?

cc @ph @blakerouse

**Comment by blakerouse** [comment]
I agree with @ph we should do them in parallel.

**Comment by nimarezainia** [comment]
> I agree with @ph we should do them in parallel.

also needs to be deterministic. We should have a limit on retry.

**Comment by AndersonQ** [comment]
I agree the agent should start applications concurrently, however making it concurrent does not solves the problem of the start up process getting stuck. Therefore I believe they should be 2 different tickets, 1 bug and 1 enhancement.

**Comment by nimarezainia** [comment]
@AndersonQ and @ph how do we want to proceed with this issue? as the enhancement i think we start in parallel and provide a time out.

the bug here seemed *manufactured* but in the future when something fails to install in the interval provided - I agree that is a bug and needs to be resolved.

**Comment by AndersonQ** [comment]
> how do we want to proceed with this issue? as the enhancement i think we start in parallel and provide a time out.

Sounds like a plan

> he bug here seemed manufactured but in the future when something fails to install in the interval provided - I agree that is a bug and needs to be resolved.

Sorry I don't follow, what do you mean by _bug here seemed **manufactured**_?

Anyway regarding failing to to install/start an application I believe the desired outcome is:
 - once it fails to install an application, mark the application as "failing", keep the agent as "configuring" on fleet-ui, move to the nest application in the list
 - once all applications are installed (either running or failing), the agent status is "healthy/unhealthy" depending on if all applications are health or not.

Ideally we won't have a infinite loop retrying to install an application. However I don't know how simple would be to skip it and keep track of it as a running application that is failing/broken. Anyway we'd need it, so we'd need to do it anyway.

> how do we want to proceed with this issue?

I'd:
 - make this ticket a bug add it to the planning. 
 - create a enhancement ticket for the concurrent start. This should be done after the bug is fixed

**Comment by ph** [comment]
I agree with your plan above @AndersonQ.

Is this following assumption correct, we have a counter in place for each application, on failure the counter is increased to a set maximun, when the maximun is reached we move it to unhealth and we stop retry to install the application. If a new agent policy is received the counter is reset to 0 and we retry again.

**Comment by AndersonQ** [comment]
I'd need to investigate the details. I imagine the counter is only for checkin, not installation attempts.

**Comment by jlind23** [comment]
I scoped it for 8.4, we will do the investigation in 8.4 release.
